File: src/main/java/io/vertigo/shiny/components/dataviz/rating/ShinyRating.java
package io.vertigo.shiny.components.dataviz.rating;

import io.vertigo.shiny.components.ShinyComponent;

import io.vertigo.shiny.ShinyType;

@ShinyType("rating")
public record ShinyRating(
		String label,
		double value,
		ShinyRatingScale scale,
		int customMaxValue, // -1 means use scale
		boolean showValue,
		boolean showPercentage,
		boolean showBox,
		String separator,
		boolean allowHalfRating) implements ShinyComponent {
}

File: banshee/src/models/dataviz/rating/ShinyRating.ts
import { ShinyComponent } from '../../ShinyComponent';
import { ShinyRatingScale, getShinyRatingMaxValue } from './ShinyRatingScale';

export class ShinyRating implements ShinyComponent {
  label: string;
  value: number;
  scale: ShinyRatingScale;
  customMaxValue: number;
  showValue: boolean;
  showPercentage: boolean;
  showBox: boolean;
  separator: string;
  allowHalfRating: boolean;
  type: string = 'rating';

  constructor(
    label: string,
    value: number,
    scale: ShinyRatingScale,
    customMaxValue: number,
    showValue: boolean,
    showPercentage: boolean,
    showBox: boolean,
    separator: string,
    allowHalfRating: boolean
  ) {
    this.label = label;
    this.value = value;
    this.scale = scale;
    this.customMaxValue = customMaxValue;
    this.showValue = showValue;
    this.showPercentage = showPercentage;
    this.showBox = showBox;
    this.separator = separator;
    this.allowHalfRating = allowHalfRating;
  }

  getMaxValue(): number {
    return this.customMaxValue !== -1 ? this.customMaxValue : getShinyRatingMaxValue(this.scale);
  }
}
